libs/ts-rest/nest/src/lib/ts-rest-nest-handler.ts

```diff
@@ -0,0 +1,365 @@
+      switch (appRouterOrRoute.method) {
+        case 'GET':
+          return Get(appRouterOrRoute.path);
+        case 'POST':
+          return Post(appRouterOrRoute.path);
+        case 'PUT':
+          return Put(appRouterOrRoute.path);
+        case 'PATCH':
+          return Patch(appRouterOrRoute.path);
+        case 'DELETE':
+          return Delete(appRouterOrRoute.path);
+      }
+    })();
+
+    decorators.push(
+      apiDecorator,
+      SetMetadata(TsRestAppRouteMetadataKey, appRouterOrRoute),
+      UseInterceptors(TsRestHandlerInterceptor)
+    );
+  }
+
+  return applyDecorators(...decorators);
+};
+
+type NestHandlerImplementation<T extends AppRouter | AppRoute> =
+  T extends AppRoute
+    ? (args: TsRestRequestShape<T>) => Promise<ServerInferResponses<T>>
+    : {
+        [K in keyof T]: T[K] extends AppRoute
+          ? (
+              args: TsRestRequestShape<T[K]>
+            ) => Promise<ServerInferResponses<T[K]>>
+          : never;
+      };
+
+/**
+ *
+ * @param contract - The contract or route to implement
+ * @param implementation - Implementation of the route or entire contract as an object
+ * @returns
+ */
+export const tsRestHandler = <T extends AppRouter | AppRoute>(
+  contract: T,
+  implementation: NestHandlerImplementation<T>
+) => implementation;
+
+/**
+ * Error you can throw to return a response from a handler
+ */
+export class TsRestException<T extends AppRoute> extends HttpException {
+  constructor(route: T, response: ServerInferResponses<T>) {
+    super(response.body as Record<string, any>, response.status);
+  }
+}
+
+export const doesUrlMatchContractPath = (
+  /**
+   * @example '/posts/:id'
+   */
+  contractPath: string,
+  /**
+   * @example '/posts/1'
+   */
+  url: string
+): boolean => {
+  const contractPathParts = contractPath.split('/');
+
+  const urlParts = url.split('/');
+
+  if (contractPathParts.length !== urlParts.length) {
+    return false;
+  }
+
+  for (let i = 0; i < contractPathParts.length; i++) {
+    const contractPathPart = contractPathParts[i];
+    const urlPart = urlParts[i];
+
+    if (contractPathPart.startsWith(':')) {
+      continue;
+    }
+
+    if (contractPathPart !== urlPart) {
+      return false;
+    }
+  }
+
+  return true;
+};
+
+@Injectable()
+export class TsRestHandlerInterceptor implements NestInterceptor {
+  constructor(private reflector: Reflector) {}
+
+  private getAppRouteFromContext(ctx: ExecutionContext) {
+    const req: Request = ctx.switchToHttp().getRequest();
+
+    const appRoute = this.reflector.get<AppRoute | undefined>(
+      TsRestAppRouteMetadataKey,
+      ctx.getHandler()
+    );
+
+    if (appRoute) {
+      return {
+        appRoute,
+        routeAlias: undefined,
+      };
+    }
+
+    /**
+     * Therefore we must be on a multi-handler so let's get the route in the router
+     */
+    const appRouter = this.reflector.get<AppRouter | undefined>(
+      TsRestAppRouterMetadataKey,
+      ctx.getHandler()
+    );
```

Comment:
I think rather than introducing two separate metadata keys, we can re-use `TsRestAppRouteMetadataKey` and pass `appRouterOrRoute` to the same metadata key in both cases. We can easily determine whether it's a multi handler or single handler by checking the value using `isAppRoute`